---
title: "ChIPseq In Bioconductor (part3) <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Center"
date: "http://rockefelleruniversity.github.io/RU_ChIPseq/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  isSlides: "no"  # change to no before any merging
---

```{r,include=FALSE}
suppressPackageStartupMessages(require(knitr))
library(TFBSTools)
library(GSEABase)
knitr::opts_chunk$set(echo = TRUE, tidy = T) 

```

```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides != "yes"){
  cat("# ChIPseq (part 3)

---
"    
  )
  
}

```


## Data

In todays session we will continue to review the Myc ChIPseq we were working on in our last sessions.

This include Myc ChIPseq for MEL and Ch12 celllines.

Information and files for the [Myc ChIPseq in MEL cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000EUA/)

Information and files for the [Myc ChIPseq in Ch12 cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000ERN/)

<!-- --- -->

<!-- # Data -->

<!-- We will be working with peak calls today, so we can download the MACS2 peak calls from the Encode website. -->

<!-- [Myc Mel Rep1](https://www.encodeproject.org/files/ENCFF363WUG/@@download/ENCFF363WUG.bed.gz) -->

<!-- [Myc Mel Rep2](https://www.encodeproject.org/files/ENCFF139JHS/@@download/ENCFF139JHS.bed.gz) -->

<!-- [Myc Ch12 Rep1](https://www.encodeproject.org/files/ENCFF160KXR/@@download/ENCFF160KXR.bed.gz) -->

<!-- [Myc Ch12 Rep2](https://www.encodeproject.org/files/ENCFF962BGJ/@@download/ENCFF962BGJ.bed.gz) -->

---

## Data

In the data directory we have provided peak calls from MACS2 following the processing steps outlined in our last session.

Peak calls for Myc in MEL and Ch12 cellines can be found in 

**data/peaks/**

* **data/peaks/Mel_1_peaks.xls**
* **data/peaks/Mel_2_peaks.xls**
* **data/peaks/Ch12_1_peaks.xls**
* **data/peaks/Ch12_1_peaks.xls**

---

## ChIP peaks in R

In our last session we reviewed how we can [identify putative transciption factor binding sites using peak calling programs such as MACS2.](https://rockefelleruniversity.github.io/RU_ChIPseq/presentations/slides/ChIPseq_In_Bioconductor2.html#38)


```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE,tidy=FALSE}
library(GenomicRanges)
macsPeaks <- "data/peaks/Mel_1_peaks.xls"
macsPeaks_DF <- read.delim(macsPeaks,comment.char="#")
macsPeaks_GR <- GRanges(seqnames=macsPeaks_DF[,"chr"],
                        IRanges(macsPeaks_DF[,"start"],macsPeaks_DF[,"end"]))
mcols(macsPeaks_GR) <- macsPeaks_DF[,c("abs_summit", "fold_enrichment")]
macsPeaks_GR[1:5,]
```


---
## Annotation of peaks to genes

Since transcription factors, as suggested in name, may regulate the transcription of their target genes, we used the **ChIPseeker package** to associate our peaks, representing potential transcription factor binding events, to their overlapping or closest mm10 genes.

```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE,tidy=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(ChIPseeker)
peakAnno <- annotatePeak(macsPeaks_GR, tssRegion=c(-1000, 1000), 
                         TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                         annoDb="org.Mm.eg.db")
```

---

## Annotation of peaks to genes

This allowed us to produce a GRanges or data.frame of our peaks and their predicted target genes.

```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE}
annotatedPeaksGR <- as.GRanges(peakAnno)
annotatedPeaksDF <- as.data.frame(peakAnno)
annotatedPeaksDF[1:2,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Gene Set Enrichment

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Gene Set Enrichment

---
"    
  )
  
}

```

## Gene Set testing

Transcription factors or epigenetic marks may act on specific sets of genes grouped by a common biological feature (shared Biological function, common regulation in RNAseq experiment etc).

A frequent step in ChIPseq analysis is to test whether common gene sets are enriched for transcription factor binding or epigenetic marks.

Sources of well curated gene sets include [GO consortium](http://geneontology.org/) (gene's function, biological process and cellular localisation), [REACTOME](http://www.reactome.org/) (Biological Pathways) and [MsigDB](http://software.broadinstitute.org/gsea/msigdb/) (Computationally and Experimentally derived).

---
## Gene Set testing for ChIPseq

Gene set enrichment testing may be performed on the sets of genes with peaks associated to them. In this example we will consider genes with peaks within 1000bp of a gene's TSS.

We will not access these database libraries directly in testing but will use other R/Bioconductor libraries which make extensive use of them.


<div align="center">
<img src="imgs/TSSPeak.png" alt="offset" height="250" width="600">
</div>

---
## Gene ontology and gene set testing

To perform gene set testing here, we will use the **clusterProfiler package**.

clusterProfiler provides multiple enrichment functions that allow for comparison of your gene list to known (e.g. GO, KEGG) or custom gene sets. Detailed information about all of the functionality within this package is available [here](http://yulab-smu.top/clusterProfiler-book/).

In this example we use all TSS sites we found to be overlapped by Myc peaks.

The peaks landing in TSS regions will be marked as "Promoter" in the **annotation** column of our annotated GRanges object.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
annotatedPeaksGR[1,]
```


---
## Gene ontology and gene set testing


We can extract the unique names of genes with peaks in their TSS by subsetting the annotated GRanges and retrieving gene names from the **geneId** column.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
annotatedPeaksGR_TSS <- annotatedPeaksGR[
  annotatedPeaksGR$annotation == "Promoter",]
genesWithPeakInTSS <- unique(annotatedPeaksGR_TSS$geneId)
genesWithPeakInTSS[1:2]
```

---
## Gene ontology and functional testing


Next we can extract all genes which are included in the TxDb object to use as our universe of genes for pathway enrichment.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T, message = F}
allGeneGR <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)
allGeneGR[1:2,]
allGeneIDs <- allGeneGR$gene_id
```

---
## Gene ontology and functional testing

Once we have our gene list and the universe of genes in the same format, we can use them in the **enrichGO** function to perform gene ontology analysis

For the **ont** argument, we can choose between the "BP", "MF", and "CC" subontologies, or "ALL" for all three.

```{r,eval=T,echo=T, message=F, warning=FALSE,tidy=T}
library(clusterProfiler)
library(org.Mm.eg.db)
GO_result <- enrichGO(gene = genesWithPeakInTSS, 
                      universe = allGeneIDs,
                      OrgDb = org.Mm.eg.db,
                      ont = "BP")

```

---
## Gene ontology and functional testing

We now have an **enrichResult** instance. From this object, we can extract a data frame of the most highly enriched gene ontology categories.
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
GO_result_df <- data.frame(GO_result)
GO_result_df[1:5, ]
```

---
## Gene ontology and functional testing

Network plots can be generated from any enrichResult object using the **enrichplot** package.

We measure similarities between the various significant gene sets and group them accordingly. The **showCategory** argument specifies how many top gene ontology hits to show. 
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T, fig.height=4, fig.width=8}
library(enrichplot)
GO_result_plot <- pairwise_termsim(GO_result)
emapplot(GO_result_plot, showCategory = 20)
```


---
## Gene ontology and functional testing

In addition to gene ontology, we can test our gene list against custom gene sets that we import as gmt files using the clusterProfiler **enricher** function. Similar to the enrichGO function, this will produce an enrichResult object that can be used for visualization. 

Here we will get gene sets from MSigDB using the **msigdbr** package. We can use this package to pull out specific gene sets, which can be seen [at this link](http://software.broadinstitute.org/gsea/msigdb/). We can also run the **msigdbr_collections** function to see the categories and subcategory codes that will be used for accessing the gene sets. 

````{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(msigdbr)
msigdbr_collections()
```

---

## Gene ontology and functional testing

From the data frame on the previous slide we can identify the category/subcategory we want, and use those in the **msigdbr** function. Here we will use "H" to access the Hallmark gene sets, and in the end we need to get a data frame where the first column contains the name of the gene sets and the second column contains the gene ID. 

````{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(msigdbr)
msig_t2g <- msigdbr(species = "Mus musculus", 
                    category = "H", 
                    subcategory = NULL)
msig_t2g <- msig_t2g[ , colnames(msig_t2g) %in% c("gs_name", "entrez_gene")]
msig_t2g[1:3, ]
```

---
## Gene ontology and functional testing

We then run the gene set enrichment, using the term to gene mapping we created as the **TERM2GENE** argument in the enricher function. 
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
hallmark <- enricher(gene = genesWithPeakInTSS, 
                     universe = allGeneIDs,
                     TERM2GENE = msig_t2g)
hallmark_df <- data.frame(hallmark)
hallmark_df[1:3, ]
```

---
## Gene ontology and functional testing

We learned about the **goseq** package during the RNAseq course, which is another functional annotation package similar to clusterProfiler,

Here we perform the same enrichment test for the MSigDB Hallmark gene sets. 

For goseq, we need a named vector of all genes (the universe) with 1s or 0s representing whether a gene had peak in TSS or not. 
We can turn a logical vector into 1 for TRUE and 0 for FALSE simply using the **as.integer** function.

```{r,eval=T,echo=T, warning=FALSE,tidy=T}
allGenesForGOseq <- as.integer(allGeneIDs %in% genesWithPeakInTSS)
names(allGenesForGOseq) <- allGeneIDs
allGenesForGOseq[1:3]
```

---
## Gene ontology and functional testing

First we must construct a **nullp** data.frame for use within **goseq** using the **nullp** function and supplying our named vector, genome to be used and gene identifier used.

The **nullp** function attempts to correct for gene length biases we may see in gene set testing. i.e. a longer gene may have more chance to have a peak within it.


```{r,include=FALSE}
library(goseq)
```

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(goseq)
pwf=nullp(allGenesForGOseq,"mm10","knownGene",plot.fit=FALSE)

```


---
## Gene ontology and functional testing

We can use the same term to gene mapping we used for clusterProfiler (though it must be converted from a tibble to data frame for goseq) to run the gene set enrichment test.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE, message = F, tidy=T}
Myc_hallMarks <- goseq(pwf,"mm10","knownGene",
                       gene2cat = data.frame(msig_t2g))

Myc_hallMarks[1:3, ]
```

---


## GREAT for GO and functional testing

In addition to a standard enrichment tests, methods have been implemented specifically for ChIPseq. Many of these tools aim to incorporate peaks distal to genes into their enrichment testing such as the popular [GREAT](http://bejerano.stanford.edu/great/public/html/splash.php) toolset.

Incorporating distal peaks by rules such as nearest gene results in some genes having a higher chance of being selected and hence some gene sets as a whole having a higher chance of having its members selected.

[GREAT](http://bejerano.stanford.edu/great/public/html/splash.php) defines regulatory regions for each individual gene and compares the proportion of peaks mapping to a gene set's regulatory regions to the proportion of the genome occupied by gene set's regulatory regions.

i.e. If a gene set's regulatory regions account for 1% of the genome then one might expect 1% of peaks to overlap these regions by chance.

---
rGREAT - R interface to GREAT server
=========================================================

We can use the GREAT Bioconductor interface available in the rGREAT package. 

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
library(rGREAT)
```
---
rGREAT for GO and functional testing
=========================================================

To submit jobs we can use our GRanges of Myc peaks and specify a genome with the **submitGreatJob** function.

This function returns a GreatJob object containing a reference to our results on the GREAT server. To review the categories of results available we can use the availableCategories function on our GreatJob object.

```{r,eval=T,echo=T, eval=T, echo=T,messages=F,message=F,warning=FALSE,tidy=T}
great_Job <- submitGreatJob(macsPeaks_GR,species="mm10",version = "3.0.0",request_interval = 1)
availableCategories(great_Job)
```
---
rGREAT for GO and functional testing
=========================================================

The results table can be retrieved using the getEnrichmentTables function and specifying which tables we wish to review.

Here we retrieve the results tables for the "Regulatory Motifs" gene sets which contains 2 separate database results.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE, message =F}
great_ResultTable = getEnrichmentTables(great_Job,
                                        category="Regulatory Motifs")
names(great_ResultTable)
```
---
rGREAT for GO and functional testin
=========================================================

Now we can review the enrichment of our genes with Myc peaks in their TSS for the "MSigDB Predicted Promoter Motifs" gene sets.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
msigProMotifs <- great_ResultTable[["MSigDB Predicted Promoter Motifs"]]
msigProMotifs[1:4,]
```

---
```{r, results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("class: inverse, center, middle

# Motifs

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("# Motifs

---
"    
  )
  
}

```

## Motifs

A common practice in transcription factor ChIPseq is to investigate the motifs enriched under peaks. 

De novo motif enrichment can be performed in R/Bioconductor but this can be very time consuming. Here we will use the MEME-ChIP suite available online to identify de novo motifs.

MEME-ChIP requires a FASTA file of sequences under peaks as input so we extract this using the **BSgenome** package.

---

## Extracting sequences under regions


First we need to load the BSgenome object for the genome we are working on, UCSC's mm10 build for the mouse genome, **BSgenome.Mmusculus.UCSC.mm10**.

```{r, echo=TRUE,include=FALSE}

library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm10)
BSgenome.Mmusculus.UCSC.mm10
```


```{r, echo=TRUE,collapse=F}

library(BSgenome)
library(BSgenome.Mmusculus.UCSC.mm10)
BSgenome.Mmusculus.UCSC.mm10
```
---

## Extracting sequences under regions


The motif for the ChIP-ed transcription factor should near the summit of a peak. MEME-ChIP will trim our peaks to a common length internally if sequences are of different length.

It is best therefore to provide a peak set resized to a common length.

```{r, echo=TRUE,collapse=F}
macsSummits_GR <- GRanges(seqnames(macsPeaks_GR),
                          IRanges(macsPeaks_GR$abs_summit,
                                  macsPeaks_GR$abs_summit),
                          score=macsPeaks_GR$fold_enrichment)
macsSummits_GR <- resize(macsSummits_GR,100,fix="center")

```

---
## Extracting sequences under regions

We now have a GRanges, centred on the summit, highest point of signal for every peak.

```{r, echo=TRUE,collapse=F}
macsSummits_GR
```

---
## Extracting sequences under regions


Once we have recentered our peaks we can use the **getSeq** function with our GRanges of resized common peaks and the BSgenome object for mm10.

The **getSeq** function returns a *DNAStringSet* object containing sequences under peaks. 

```{r, echo=TRUE,collapse=F}
peaksSequences <- getSeq(BSgenome.Mmusculus.UCSC.mm10,
                         macsSummits_GR)
names(peaksSequences) <- paste0(seqnames(macsSummits_GR),":",
                                         start(macsSummits_GR),
                                         "-",
                                         end(macsSummits_GR))

peaksSequences[1:2,]
```
---
## Writing to FASTA file


The *writeXStringSet* function allows the user to write DNA/RNA/AA(aminoacid)StringSet objects out to file. 

By default the *writeXStringSet* function writes the sequence information in FASTA format (as required for MEME-ChIP).

```{r, echo=TRUE,collapse=F}
writeXStringSet(peaksSequences,file="mycMel_rep1.fa")

```

---
## MEME-ChIP

Now the file "mycMel_rep1.fa" contains sequences around the geometric center of peaks suitable for Motif analysis in MEME-ChIP. 

In your own work you will typically run this from your own laptop with MEME installed locally but today we will upload our generated FASTA file to their [web portal](http://meme-suite.org/tools/meme-chip). 

Follow instructions [here](http://meme-suite.org/doc/download.html) to install MEME localy.

Results files from MEME-ChIP can be found [here](http://rockefelleruniversity.github.io/myc_Meme_Example/meme-chip.html)

---
## Parsing back FIMO results

We can retrieve the locations of Myc motifs identified in MEME-ChIP from the FIMO output.

FIMO reports Myc motif locations as a GFF3 file which we should be able to vizualise in IGV. 
Sadly, this GFF file's naming conventions cause only a fraction of motifs to be reported.

<div align="center">
<img src="imgs/fimoBad.png" alt="offset" height="300" width="600">
</div>

---
## FIMO to R

Fortunately we can parse our motif's GFF file into R and address this using the **import** function in  the **rtracklayer** package.

```{r, echo=TRUE,collapse=F,eval=FALSE}
library(rtracklayer)
motifGFF <- import("~/Downloads/fimo.gff")
```

---
## FIMO to valid GFF3

We can give the sequences some more sensible names and export the GFF to file to visualise in IGV.

```{r, echo=TRUE,collapse=F,eval=FALSE}
motifGFF$Name <- paste0(seqnames(motifGFF),":",
                        start(motifGFF),"-",end(motifGFF))
motifGFF$ID <- paste0(seqnames(motifGFF),":",
                      start(motifGFF),"-",end(motifGFF))
export.gff3(motifGFF,con="~/Downloads/fimoUpdated.gff")
```

<div align="center">
<img src="imgs/fimoGood.png" alt="offset" height="250" width="600">
</div>

---
## Scanning for known motifs 

We saw previously we can scan sequences using some of the Biostrings functionality **matchPattern**.

Often with ChIPseq we may know the motif we are looking for or we can use a set of known motifs from a database such as a [JASPAR](http://jaspar.genereg.net).

We can access to JASPAR using the JASPAR2020 bioconductor library.

```{r, echo=TRUE,collapse=F,eval=TRUE}
library(JASPAR2020)
JASPAR2020
```

---
## Get motifs from JASPAR with TFBStools

We can access the model for the our motif of interest using the **TFBSTools** package and its **getMatrixByName** function.

```{r, echo=TRUE,collapse=F,eval=TRUE}
library(TFBSTools)
pfm <- getMatrixByName(JASPAR2020, 
                       name="MYC")
pfm
```

---
## Motif scanning with motifmatchr

With this PWM we can use the **motifmatchr** package to scan our summits for the Myc motif and return the positions of the motifs.

We will need to provide our PWM, GRanges to scan within and BSGenome object to extract sequence from. 

We also set the **out** paramter to positions for this instance.

```{r, echo=TRUE,collapse=F,eval=TRUE}
library(motifmatchr)
MycMotifs <- matchMotifs(pfm,
                         macsSummits_GR,BSgenome.Mmusculus.UCSC.mm10, 
                         out = "positions")
MycMotifs
```

---
## Exporting motif matches

We can export the Myc motif positions within peaks for use later in IGV or for metaplot vizualisation.


```{r}
export.bed(MycMotifs[[1]],con = "MycMotifs.bed")
```

<!-- --- -->
<!-- # High confidence peaks -->

<!-- As we discussed in our last session, ChIPseq data will often contain artefact signal and noise. Following the work by the Encode consortium, it is becoming more standard practice to have replicate samples for ChIPseq. -->

<!-- One approach to selecting reproducible peaks in ChIPseq is to identify peaks which are present in the majority of replicates. -->

<!-- We can first read in the other replicate dataset for MEL Myc ChIPseq below. -->

<!-- ```{r, echo=TRUE,collapse=F} -->

<!-- library(GenomicRanges) -->
<!-- library(TxDb.Mmusculus.UCSC.mm10.knownGene) -->
<!-- library(ChIPseeker) -->
<!-- macsPeaks <- "data/peaks/Mel_2_peaks.xls" -->
<!-- macsPeaks_DF2 <- read.delim(macsPeaks,comment.char="#") -->
<!-- macsPeaks_GR2 <- GRanges( -->
<!--  seqnames=macsPeaks_DF2[,"chr"], -->
<!--  IRanges(macsPeaks_DF2[,"start"],macsPeaks_DF2[,"end"]) -->
<!-- ) -->
<!-- mcols(macsPeaks_GR2) <- macsPeaks_DF2[,c("abs_summit", "fold_enrichment")] -->
<!-- ``` -->

<!-- --- -->
<!-- # Manipulating Peak Sets - Finding Common peaks -->


<!-- When looking at peaks which occur in both samples it is clear that the number of peaks in first replicate overlapping those in second is different from number of second replicate peaks overlapping first. -->

<!-- This is because 2 peaks from one replicate may overlap 1 peak in the other replicate. -->

<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- firstANDsecondPeakSets <- macsPeaks_GR[macsPeaks_GR %over% macsPeaks_GR2] -->
<!-- secondANDfirstPeakSets <- macsPeaks_GR2[macsPeaks_GR2 %over% macsPeaks_GR] -->

<!-- length(firstANDsecondPeakSets) -->
<!-- length(secondANDfirstPeakSets) -->
<!-- ``` -->

<!-- --- -->
<!-- ![alt text](imgs/oneToMany.png) -->

<!-- --- -->
<!-- # Manipulating Peak Sets - Finding Common peaks -->


<!-- A common step with finding overlapping transcription factor peaks is to reduce peaksets to single non-overlapping peakset before interrogating whether a peak occurred in a sample. -->

<!-- This allows for a single peak set to be used as a consensus peakset between replicates. -->


<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- allPeaks <- c(macsPeaks_GR,macsPeaks_GR2) -->
<!-- allPeaksReduced <- reduce(allPeaks) -->
<!-- length(allPeaks) -->
<!-- length(allPeaksReduced) -->
<!-- ``` -->

<!-- --- -->
<!-- ![alt text](imgs/mel_Flattened.png) -->


<!-- --- -->

<!-- Now we can use a logical expression to subset our reduced/flattened peak set to those overlapping peaks in both the first and second replicate. -->

<!-- ```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE} -->
<!-- commonPeaks <- allPeaksReduced[allPeaksReduced %over% macsPeaks_GR  -->
<!--                                & allPeaksReduced %over% macsPeaks_GR2] -->
<!-- length(commonPeaks) -->
<!-- ``` -->

<!-- --- -->
<!-- ![alt text](imgs/Ch12_highcon.png) -->

---
## Time for an exercise!

Exercise on ChIPseq data can be found [here](../../exercises/exercises/chipseq_part3_exercise.html)

---
## Answers to exercise

Answers can be found [here](../../exercises/answers/chipseq_part3_answers.html)



