---
title: "ChIP-seq In Bioconductor part2"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/RU_ChIPseq/"
output: 
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style


---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


# Data

In todays session we will continue to review the Myc ChIPseq we were working on in our last session.

This include Myc ChIP-seq for MEL and Ch12 celllines as well as their input controls.

Information and files for the [Myc ChIPseq in MEL cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000EUA/)

Information and files for the [Myc ChIPseq in Ch12 cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000ERN/)

Input control can be found for [MEL cell line can be found here](https://www.encodeproject.org/experiments/ENCSR000EUA/)

Input control can be found for [Ch12 cell line can be found here.](https://www.encodeproject.org/experiments/ENCSR000ERN/)




---
# Quality Control.


ChIP-seq has many sources of potential noise including 
* Varying efficiency of antibodies
* Non-specific binding
* Library complexity
* ChIP artefacts and background.

Many of these sources of noise can be assessed using some now well-established methodology.

---

# Quality Control. - Some references


For some discussions:

* Encode quality metrics.

[Large-scale quality analysis of published ChIP-seq data. Marinov GK, Kundaje A, Park PJ, Wold BJ. G3 (Bethesda). 2014 Feb 19;4(2)](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3931556/)

* Overestimation of artefact duplicates in ChIPseq.

[Systematic evaluation of factors influencing ChIP-seq fidelity.Nat Methods. Chen Y, Negre N, Li Q, Mieczkowska JO, Slattery M, Liu T, Zhang Y, Kim TK, He HH, Zieba J, Ruan Y, Bickel PJ, Myers RM, Wold BJ, White KP, Lieb JD, Liu XS. 2012 Jun;9(6)](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3477507/)


* When and what QC is useful.

[Impact of artifact removal on ChIP quality metrics in ChIP-seq and ChIP-exo data.Front Genet. 2014 Apr 10;5:75.Carroll TS, Liang Z, Salama R, Stark R, de Santiago I](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3989762/)

---
# Quality Control - Always have an appropriate input.


* Input samples are typically made from fragmented DNA prior to IP enrichment.

* Allows for control of artefact regions which occur across samples.

* NEVER run ChIP-seq without considering which input to use.

e.g. When using tumour samples for ChIP-seq, it is important to have matched input samples. 
Differing conditions of same tissue may share common input. 
---
# Quality Control - Quality metrics for ChIP-seq.


The [**ChIPQC package**](https://bioconductor.org/packages/release/bioc/html/ChIPQC.html) wraps some of the metrics into a Bioconductor package and takes care to measure these metrics under the appropriate condition. 

To run a single sample we can use the **ChIPQCsample()** function, the relevant **unfiltered** BAM file and we are recommended to supply a **blacklist** as a BED file or GRanges and Genome name.

You can find a Blacklist for most genomes at [Anshul Kundaje's site](https://sites.google.com/site/anshulkundaje/projects/blacklists) or directly from the [Encode websites](https://www.encodeproject.org/annotations/ENCSR636HFF/)

```{r mycQCdwdwshowL,include=FALSE}
library(ChIPQC)
```


```{r,eval=F} 
QCresult <- ChIPQCsample(reads="/pathTo/myChIPreads.bam",
                         genome="mm10",
                         blacklist = "/pathTo/mm10_Blacklist.bed")
```

---
# Quality control

We download the blacklist for mm10 from [Encode ](https://www.encodeproject.org/files/ENCFF547MET/@@download/ENCFF547MET.bed.gz)

We can then provide an initial analysis of our ChIP-seq samples quality using the **ChIPQCsample()** function from the [**ChIPQC** package.](http://bioconductor.org/packages/stats/bioc/ChIPQC/).

Here we evalute the quality of samples we aligned with Rsubread last week. The returned object is a **ChIPQCsample** object.

```{r mycQC,cache=TRUE,eval=FALSE}
library(ChIPQC)
chipqc_MycMel_rep1 <- ChIPQCsample("SR_Myc_Mel_rep1.bam",
                         annotation = "mm10",
                         blacklist = toBlkList,
                         chromosomes = paste0(c("chr",1:10)))
class(chipqc_MycMel_rep1)
```

```{r mycQCshowLa,echo=FALSE,eval=TRUE}
library(ChIPQC)
load(file="rep1.RData")
class(chipqc_MycMel_rep1)
```

---
# Quality control

We can display our **ChIPQCsample** object which will show a basic summary of our ChIP-seq quality.

```{r mycQCshowL,include=FALSE}
library(ChIPQC)
load(file="rep1.RData")
```

```{r mycQCshow,eval=TRUE}
chipqc_MycMel_rep1
```

---
# Quality control

It is best to review ChIP-seq quality versus your input controls and other Myc samples which we are using (or even external data if you do not have your own).

This will allow us to identify expected patterns of ChIP-seq enrichment in our samples versus controls as well as spot any outlier samples by these metrics.

We can run **ChIPQCsample()** on all our samples of interest using an **lapply**.

```{r mycQCshowd2,cache=TRUE,eval=FALSE,include=FALSE}
setwd("~/Projects/Results/chipseq/testRun/BAMs/")
bamsToQC <- c("Sorted_Myc_Ch12_1.bam","Sorted_Myc_Ch12_2.bam",
             "Sorted_Myc_MEL_1.bam","Sorted_Myc_MEL_2.bam",
             "Sorted_Input_MEL.bam","Sorted_Input_Ch12.bam")
myQC <- bplapply(bamsToQC,ChIPQCsample,
        annotation = "mm10",
        blacklist = toBlkList,
        chromosomes = paste0("chr",1:10))
names(myQC) <- bamsToQC
```

```{r mycQCshow2,cache=TRUE,eval=FALSE}
bamsToQC <- c("Sorted_Myc_Ch12_1.bam","Sorted_Myc_Ch12_2.bam",
             "Sorted_Myc_MEL_1.bam","Sorted_Myc_MEL_2.bam",
             "Sorted_Input_MEL.bam","Sorted_Input_Ch12.bam")
myQC <- bplapply(bamsToQC,ChIPQCsample,
        annotation = "mm10",
        blacklist = toBlkList,
        chromosomes = paste0("chr",1:10))
names(myQC) <- bamsToQC
```

---
# Quality control

All ChIPQC functions can work with a named list of **ChIPQCsample** objects to aggregate scores into table as well as plots.

Here we use the **QCmetrics()** function to ive an overview of quality metrics. 

```{r qcmetricsA,include=FALSE}
load(file="myQCnoPeaks.RData")
```

```{r qcmetrics,cache=FALSE,eval=TRUE}
QCmetrics(myQC)
```

---

# Quality Control (Assessing fragment length)


The prediction of fragment length is an essential part of ChIP-seq affecting peaks calling, summit identification and coverage profiles. 

The use of cross-correlation or cross-coverage allows for an assessment of reads clustering by strand and so a measure of quality. 

---
# Quality Control (Assessing fragment length)

<div align="center">
<img src="imgs/ChIP-seq_biology_slides.png" alt="offset" height="400" width="600">
</div>

---

# Quality Control (Assessing fragment length)

* In ChIP-seq typically short single end reads of dsDNA.

* **dsDNA single end sequencing means**
+ 5' of fragment will be sequenced on "+" strand
+ 3' of fragment end will be on "-" strand.

* **Although we only have partial sequence of strand, with predicted fragment length we can predicted the whole fragment**
+ "+" reads should extend only in positive direction 
+ "-" reads only in negative

---

# Quality Control (Assessing fragment length)

<div align="center">
<img src="imgs/pileup.png" alt="offset" height="500" width="400">
</div>

---

# Quality Control (Assessing fragment length)

<div align="center">
<img src="imgs/offset.jpg" alt="offset" height="500" width="400">
</div>

---

# Quality Control (Assessing fragment length)

<div align="center">
<img src="imgs/shifts.gif" alt="offset" height="200" width="700">
</div>
<div align="center">
<img src="imgs/cor.gif" alt="offset" height="300" width="700">
</div>

---

# Quality Control (Assessing fragment length)

<div align="center">
<img src="imgs/shifts.jpg" alt="offset" height="500" width="400">
</div>

---

# Quality Control (Assessing fragment length)
The **plotCC** function can be used to plot our cross-coverage profiles

The **plotCC()** function accepts our list of ChIPQC sample objects and a **facetBy** argument to allow us to group our cross-coverage profiles.

```{r qcmetridedecs,cache=FALSE,eval=TRUE,fig.width=6,fig.height=4}
plotCC(myQC,facetBy = "Sample")
```

---

# Quality Control (Assessing fragment length)

We can include additional metadata to allow us to group our plot in different ways.

We can include metadata as a data.frame where the first column is our sample names.


```{r qcmetridecs,cache=FALSE,eval=TRUE}
myMeta <- data.frame(Sample= names(myQC),
                     Tissue=c("Ch12","Ch12","MEL","MEL","MEL","Ch12"),
                     Antibody=c(rep("Myc",4),rep("Input",2)))
myMeta
```

---
# Quality Control (Assessing fragment length)

We can now include our metadata to the **addMetaData** argument which will allow us to **facetBy** the supplied metadata columns.

Additionally here we use the **colourBy** parameter to add colour to antibody groups.

```{r qcmetricsede,cache=FALSE,eval=TRUE,fig.width=6,fig.height=3}
plotCC(myQC,facetBy = "Tissue",addMetaData = myMeta,
       colourBy="Antibody")
```
---
# Quality Control (Assessing fragment length)

All plots in ChIPQC are in fact built in ggplot2 so we can edit and update our plot like all ggplot objects.

```{r qcmetricsrf,cache=FALSE,eval=TRUE,fig.width=6,fig.height=3}
plotCC(myQC,facetBy = "Tissue",addMetaData = myMeta,
       colourBy="Antibody")+theme_bw()+
  ggtitle("ChIPQC results")
```


---
# Quality Control - Blacklists and SSD.


ChIP-seq will often show the presence of common artefacts such as ultra-high signal regions or **Blacklists**. Such regions can confound peak calling, fragment length estimation and QC metrics.

SSD is a measure of standard deviation of signal across the genome with higher scores reflecting significant pile-up of reads. SSD can therefore be used to assess both the extent of ultra high signals and the signal following removal of these blacklisted regions.

For a note on known Blacklisted regions and on associated resources.
* [Blacklisted Regions](http://mrccsc.github.io/analysisbeginings.html#/35)

For a note on SSD
* [SSD and Signal Pileup](http://mrccsc.github.io/analysisbeginings.html#/36).

---
# Quality Control - Blacklist???

<div align="center">
<img src="imgs/blacklist.png" alt="offset" height="500" width="400">
</div>

---
# Quality Control - Blacklist affects many metrics.


<div align="center">
<img src="imgs/blacklistsAffects.jpg" alt="offset" height="400" width="400">
</div>

---
# Quality Control - Blacklist affects many metrics.


<div align="center">
<img src="imgs/ssdAndBlacklist.png" alt="offset" height="400" width="300">
</div>


---
# Quality Control - Standardised Standard Deviation.


ChIPQC calculates SSD before and after removing signal coming from Blacklisted regions.

The **plotSSD()** function plots samples's pre-blacklisting score in **red** and post-blacklisting score in **blue**.

Higher scores for pre-blacklisted SSD can suggest a strong background signal in blacklisted regions for that sample.

```{r fig.width=6,fig.height=2,warning=FALSE,message=FALSE}
plotSSD(myQC)+xlim(0,5)
```

---
# Quality Control - Standardised Standard Deviation.


Since SSD score is strongly affected by blacklisting it may be necessary to change the axis to see any differences between samples for post-blacklisting scores.

Higher post-blacklisted SSD scores reflect samples with stronger peak signal.

```{r fig.width=6,fig.height=3,warning=FALSE,message=FALSE}
plotSSD(myQC)+xlim(0.2,0.4)
```

---
# Quality Control - Library complexity

A potential source of noise in ChIPseq is overamplification of the ChIP-seq library during a PCR step. This can lead to large number of duplicate reads which may confound peak calling.

![](imgs/mappable.png)

---
# Quality Control - Duplication.


We should compare our duplication rate across samples to identify any sample experiencing overamplification and so potential of a lower complexity.

The **flagtagcounts()** function reports can report the number of duplicates and total mapped reads and so from there we can calculate our duplication rate.

```{r fig.width=6,fig.height=3,warning=FALSE,message=FALSE}
myFlags <- flagtagcounts(myQC)
myFlags["DuplicateByChIPQC",]/myFlags["Mapped",]
```


---
# Quality Control - Enrichment for reads across genes.

We can also use ChIPQC to review where our distribution of reads across gene features using the **plotRegi()** function.

Here we expect ChIPseq signal to be stronger in 5'UTRs and promoters when compared to input samples.

```{r warning=FALSE,message=FALSE,fig.width=8,fig.height=4}
p <- plotRegi(myQC)
```

---
# Quality Control - Enrichment for reads across genes.

```{r warning=FALSE,fig.width=12,fig.height=6}
p
```


---
# Peak calling

To identify regions of Myc transcription factor binding we can use a **peak caller**.

Although peak caller are available in R (including our own), the most popular and widely used peak called remains **MACS2**.

MACS2 calls peaks in a few simple steps.

* Predict fragment length.
* Shift reads to centre of predicted fragment.
* Scan across genome and identify regions enriched compare to control samples using count based statistic.

---
# Peak calling (Installing MACS2)

![testing](imgs/bioconda.jpg){width=50%}

* Install conda with **[Anaconda](https://anaconda.org/)** or **[Miniconda](https://conda.io/miniconda.html)**


* Set bioconda channels accroding to the *[construction](https://bioconda.github.io/index.html#set-up-channels)*

* Install packages e.g. *[MACS2](https://bioconda.github.io/recipes/macs2/README.html)*



---
# Peak calling (Running MACS2)

To run MACS2 to we simply need to supply.

* A BAM file to find enriched regions in. (specified after **-t**)
* A Name for peak calls (specified after **–name**).
* An output folder to write peaks into (specified after **–outdir**).
* Optionally, but highly recommended, we can identify a control to compare to (specified after **–c**).

```{r , engine='bash', eval=FALSE}
macs2 callpeak -t Sorted_Myc_MEL_1.bam
               –name Mel_Rep1 
               –-outdir PeakDirectory
               -c Sorted_Input_MEL.bam 

```
---
# Peak calling from R

We can still run our MACS2 from the comfort of R using the **system()** function.

We simply need to build our MACS2 peak call command in R and pass to the **system()** function. 

```{r fig.height=5, fig.width=15,eval=FALSE}
myChIP <- "Sorted_Myc_MEL_1.bam"
myControl <- "Sorted_Input_MEL.bam"

macsCommand <- paste0("macs2 callpeak -t ", myChIP,
                      " -n ", "Mel_Rep1",
                      " –-outdir ","PeakDirectory",
                      " -c ", myControl)
system(macsCommand)
```



---
# Working with Peaks


MACS peak calls can be found in our specied output directory with the suffix and extension "_peaks.xls".

MACS peaks come as a tab seperated file thinly disguised as a ".xls".

```{r,eval=T,echo=F,  warning=FALSE,collapse=T}
macsPeaks <- "../Slides/data/Mel1_peaks.xls"
macsPeaks_DF <- read.delim(macsPeaks,sep="",comment="#")
macsPeaks_DF[1:3,]
```

---
#Working with Peaks


In addition to the genomic coordinates of peaks, these files contain useful information on the samples, parameters and version used for peak calling at the top.

```{r,eval=T,echo=F,  warning=FALSE,collapse=T}
macsPeaks_DF <- read.delim(macsPeaks,comment="",stringsAsFactors = F)
macsPeaks_DF[1:8,]
```

---

# Working with Peaks - Importing peaks


We can import peak files therefore using read.delim function. Note we have set *comment.char* argument to **#** to exclude additional information on peak calling parameters stored within the MACS peak file.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
macsPeaks_DF <- read.delim(macsPeaks,comment.char="#")
macsPeaks_DF[1:2,]
```

---
# Working with Peaks - Importing peaks


Now we have the information in a table we can create a GRanges object.

GRanges objects are made of chromosome names and intervals stored as IRanges.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
library(GenomicRanges)
macsPeaks_GR <- GRanges(
 seqnames=macsPeaks_DF[,"chr"],
 IRanges(macsPeaks_DF[,"start"],
         macsPeaks_DF[,"end"]
 )
)
macsPeaks_GR
```

---
# Working with Peaks - Peaks as GRanges


As we have seen before elements in GRanges can accessed and set using various GRanges functions.
Here we can deconstruct our object back to contig names and interval ranges.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
seqnames(macsPeaks_GR)
ranges(macsPeaks_GR)
```

---
# Working with Peaks - Peaks as GRanges


GRanges objects may have metadata attached. Here we attach some useful information on our peaks including the summit position and the fold enrichment over input.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
mcols(macsPeaks_GR) <- macsPeaks_DF[,c("abs_summit", "fold_enrichment")]
macsPeaks_GR
```


---
# Working with Peaks - Filter peaks in blacklisted regions.

We will want to remove any peaks overlapping blacklisted regions prior to any downstream analysis. 
We can do this using simple overlapping with GRanges objects.


```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE}
library(rtracklayer)
blkList <- import.bed("mm10-blacklist.bed")
macsPeaks_GR <- macsPeaks_GR[!macsPeaks_GR %over% blkList] 
```

---

# Annotation of peaks to genes

So far we have been working with ChIP-seq peaks corresponding to transcription factor binding. Transcription factors, as implied in the name, can affect the expression of their target genes.

The target of transcription factor is hard to assertain from ChIP-seq data alone and so often we will annotate peaks to genes by a simple set of rules:-

Peaks are typically annotated to a gene if
* They overlap the gene.
* The gene is the closest (and within a minimum distance.)



---

# Peak annotation


A useful package for annotation of peaks to genes is **ChIPseeker**. 

By using pre-defined annotation in the from of a **TXDB** object for mouse (mm9 genome), ChIPseeker will provide us with an overview of where peaks land in the gene and distance to TSS sites.

First load the libraries we require for the next part.

```{r,include=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(GenomeInfoDb)
library(ChIPseeker)

```

```{r,eval=F,echo=T, eval=T, echo=T, warning=FALSE,tidy=T,message=FALSE}
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(GenomeInfoDb)
library(ChIPseeker)

```

---

# Peak annotation


The annotatePeak function accepts a GRanges object of the regions to annotate, a TXDB object for gene locations and a database object name to retrieve gene names from.


```{r,eval=T,echo=T, message=FALSE,messages=FALSE, eval=T, echo=T, warning=FALSE}
peakAnno <- annotatePeak(macsPeaks_GR, tssRegion=c(-500, 500), 
                         TxDb=TxDb.Mmusculus.UCSC.mm10.knownGene, 
                         annoDb="org.Mm.eg.db")
class(peakAnno)
```
---
# Peak annotation


The result is a csAnno object containing annotation for peaks and overall annotation statistics.

```{r,eval=T,echo=T, message=F,messages=F, eval=T, echo=T, warning=FALSE,tidy=T}
peakAnno
```

---
#Peak annotation


The csAnno object contains the information on annotation of individual peaks to genes.

To extract this from the csAnno object the ChIPseeker functions *as.GRanges* or *as.data.frame* can be used to produce the respective object with peaks and their associated genes.

```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
peakAnno_GR <- as.GRanges(peakAnno)
peakAnno_DF <- as.data.frame(peakAnno)
```

---
#Peak annotation
```{r,eval=T,echo=T, eval=T, echo=T, warning=FALSE,tidy=T}
peakAnno_GR[1,]
```

---
# Visualising peak annotation

Now we have the annotated peaks from ChIPseeker we can use some of ChIPseeker's plotting functions to display distribution of peaks in gene features. Here we use the **plotAnnoBar** function to plot this as a bar chart but  **plotAnnoPie** would produce a similar plot as a pie chart.


```{r, eval=T, echo=T, fig.height=5, fig.width=15, warning=FALSE, tidy=T}
plotAnnoBar(peakAnno)
```

---
# Visualising peak annotation

Similarly we can plot the distribution of peaks around TSS sites.


```{r,eval=T,echo=T, eval=F, echo=T, warning=FALSE,fig.height=5, fig.width=15,tidy=T}
plotDistToTSS(peakAnno)
```

![](imgs/TSS1.png)

---
# Visualising peak annotation

ChIPseeker can also offer a succinct plot to describe the overlap between annotations.


```{r, eval=T, echo=T, fig.height=5, fig.width=15, warning=FALSE, tidy=T}
upsetplot(peakAnno, vennpie=F)
```


---

# Time for an exercise.

[Link_to_exercises](../../Exercises/chipseq_part2_exercises.html)

[Link_to_answers](../../Answers/chipseq_part2_answers.html)


