<!DOCTYPE html>
<html>
  <head>
    <title>ChIP-seq In Bioconductor</title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# ChIP-seq In Bioconductor
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/RU_ChIPseq/" class="uri">http://rockefelleruniversity.github.io/RU_ChIPseq/</a>

---






# ChIP-seq introduction 

Chromatin precipitation followed by deep sequencing (**ChIP-seq**) is a well established technique which allows for the genome wide identification of transcription factor binding sites and epigenetic marks. 

&lt;div align="center"&gt;
&lt;img src="imgs/chipOverview1.png" alt="igv" height="400" width="600"&gt;
&lt;/div&gt;

---
# ChIP-seq introduction 

.pull-left[
&lt;div align="center"&gt;
&lt;img src="imgs/chipOverview2.png" alt="igv" height="500" width="300"&gt;
&lt;/div&gt;
]

.pull-right[

* Cross-linked and protein bound DNA.
* Enrichment by antibody for specific protein or DNA state.
* End repair, A-tailed and illumina adaptors added.
* Fragments sequenced from one/either end.
]
---
# Raw Sequence data.

Our raw ChIP-seq sequencing data will be in FASTQ format.

&lt;div align="center"&gt;
&lt;img src="imgs/fastq.png" alt="igv" height="400" width="600"&gt;
&lt;/div&gt;


---
# Raw ChIPseq Sequence data.

In this ChIP-seq workshop we will be investigating the genome wide binding patterns of the transcription factor Myc in mouse MEL and Ch12 cell lines.

We can retrieve the raw sequencing data from Encode website.

Here we download the sequencing data for the Myc ChIP-seq from the Mouse MEL cell line[, sample **ENCSR000EUA** (replicate 1), using the Encode portal.](https://www.encodeproject.org/experiments/ENCSR000EUA/)

The direct link to the raw sequecing reads in FastQ format can be found [here.](https://www.encodeproject.org/files/ENCFF001NQP/@@download/ENCFF001NQP.fastq.gz)

Download the fastQ for the other Myc MEL replicate from [sample ENCSR000EUA](https://www.encodeproject.org/experiments/ENCSR000EUA/). Direct link is [here](https://www.encodeproject.org/files/ENCFF001NQQ/@@download/ENCFF001NQQ.fastq.gz).


---

# Working with raw ChIPseq sequence data.

Once we have downloaded the raw fastQ data we can use the [ShortRead package](https://bioconductor.org/packages/release/bioc/html/ShortRead.html) to review our sequence data quality.

We have reviewed how to work with raw sequencing data in the [**FastQ in Bioconductor** session.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#1)

First we load the [ShortRead library.](https://bioconductor.org/packages/release/bioc/html/ShortRead.html)





```r
library(ShortRead)
```

---

# Working with raw ChIPseq sequence data.

First we will review the raw sequencing reads using functions in the [ShortRead package.](https://bioconductor.org/packages/release/bioc/html/ShortRead.html)

We however do not need to review all reads in the file to can gain an understanding of data quality. We can simply review a subsample of the reads and save ourselves some time and memory.

Note when we subsample we retrieve random reads from across the entire fastQ file. This is important as fastQ files are often ordered by their position on the sequencer.

---

# Working with raw ChIPseq sequence data.

We can subsample from a fastQ file using functions in **ShortRead** package.

Here we use the [**FastqSampler** and **yield** function](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#39) to randomly sample a defined number of reads from a fastQ file. Here we subsample 1 million reads.


```r
fqSample &lt;- FastqSampler("~/Downloads/ENCFF001NQP.fastq.gz",n=10^6)
fastq &lt;- yield(fqSample)
```

---

# Working with raw ChIPseq sequence data.

The resulting object is a [ShortReadQ object](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#9) showing information on the number of cycles, base pairs in reads, and number of reads in memory.


```r
fastq
```

```
## class: ShortReadQ
## length: 1000000 reads; width: 36 cycles
```

---

# Working with raw ChIPseq sequence data.

If we wished, we can assess information from the fastQ file using our [familiar accessor functions.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#13)

* **sread()** - Retrieve sequence of reads.
* **quality()** - Retrieve quality of reads as ASCI scores.
* **id()** - Retrieve IDs of reads.




```r
readSequences &lt;- sread(fastq)
readQuality &lt;- quality(fastq)
readIDs &lt;- id(fastq)
readSequences
```

```
##   A DNAStringSet instance of length 1000000
##           width seq
##       [1]    36 ATATATAATGCTTTGCATANGAAAGATCCGTCACAT
##       [2]    36 ATTTCAAAATTTTCCACCCTCTTTAGTTTTCCTAGC
##       [3]    36 GGAGGGCTGCTCCGCAGAACCGGAGCATCTAAATTT
##       [4]    36 AGGGACAATCCACAGCATACTAATTGTTGGGTTAAT
##       [5]    36 GATCGGAATAGCGGGTATGTCGGTTACTGGTTGTTT
##       ...   ... ...
##  [999996]    36 AATATATCTTTAAATTTCAATGAGTCAGGTTTTCCT
##  [999997]    36 ATCATCTTACCTTTATGTACATGCATTTAAAATTAT
##  [999998]    36 ACTGGTTTGACATTGAAAGGGAAATGATTCATTTTG
##  [999999]    36 GGTGGAACTTAAGTCCTGGGCTTGTTTGCTTTCTTT
## [1000000]    36 CTCTTTGCCAGCCAGTGATAGGCTAGGAACTTCAGG
```

---

# Quality with raw ChIPseq sequence data.

We can check some simple quality metrics for our subsampled fastQ data.

First we can review the overall reads' quality scores.

We use the [**alphabetScore()** function with our read's qualitys](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#26) to retrieve the sum quality for every read from our subsample.




```r
readQuality &lt;- quality(fastq)
readQualities &lt;- alphabetScore(readQuality)
readQualities[1:10]
```

```
##  [1]  101 1166 1042 1184  669 1075  699  394 1139 1083
```

---

# Quality with raw ChIPseq sequence data.

We can then produce a histogram of quality scores to get a better understanding of the distribution of scores.


```r
library(ggplot2)
toPlot &lt;- data.frame(ReadQ=readQualities)
ggplot(toPlot,aes(x=ReadQ))+geom_histogram()+theme_minimal()
```

```
## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

![](ChIPseq_In_Bioconductor_files/figure-html/mycRep1ReadsQScoresPlot-1.png)&lt;!-- --&gt;


---

# Base frequency with raw ChIPseq sequence data.

We can review the occurrence of DNA bases within reads and well as the occurrence of DNA bases across sequencing cycles using the [**alphabetFrequency()**](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#16) and [**alphabetByCycle()**](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#26) functions respectively.

Here we check the overall frequency of **A, G, C, T and N (unknown bases)** in our sequence reads.


```r
readSequences &lt;- sread(fastq)
readSequences_AlpFreq &lt;- alphabetFrequency(readSequences)
readSequences_AlpFreq[1:3,]
```

```
##       A  C  G  T M R W S Y K V H D B N - + .
## [1,] 13  6  5 11 0 0 0 0 0 0 0 0 0 0 1 0 0 0
## [2,]  8 10  2 16 0 0 0 0 0 0 0 0 0 0 0 0 0 0
## [3,]  9  9 11  7 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```


---

# Base frequency  with raw ChIPseq sequence data.

Once we have the  frequency of DNA bases in our sequence reads we can retrieve the sum across all reads.


```r
summed__AlpFreq  &lt;- colSums(readSequences_AlpFreq)
summed__AlpFreq[c("A","C","G","T","N")]
```

```
##        A        C        G        T        N 
## 10035623  7842724  7648164 10096682   376807
```

---

# Assessment by cycle  with raw ChIPseq sequence data.

We can review DNA base occurrence by cycle using the [**alphabetByCycle()** function.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#26)


```r
readSequences_AlpbyCycle &lt;- alphabetByCycle(readSequences)
readSequences_AlpbyCycle[1:4,1:10]
```

```
##         cycle
## alphabet   [,1]   [,2]   [,3]   [,4]   [,5]   [,6]   [,7]   [,8]   [,9]
##        A 306465 329104 283423 276630 277071 276360 278836 282486 271829
##        C 179608 174653 206137 223713 210286 221216 214314 217707 220437
##        G 242964 245564 226757 211957 213186 214652 210585 209291 213414
##        T 266782 247203 278641 278081 289735 278436 286920 279331 283428
##         cycle
## alphabet  [,10]
##        A 279283
##        C 222588
##        G 209385
##        T 277852
```

---

# Assessment by cycle  with raw ChIPseq sequence data.

We often plot this to visualise the base occurrence over cycles to observe any bias.
First we arrange the base frequency into a data frame.


```r
AFreq &lt;- readSequences_AlpbyCycle["A",]
CFreq &lt;- readSequences_AlpbyCycle["C",]
GFreq &lt;- readSequences_AlpbyCycle["G",]
TFreq &lt;- readSequences_AlpbyCycle["T",]
toPlot &lt;- data.frame(Count=c(AFreq,CFreq,GFreq,TFreq),
                     Cycle=rep(1:36,4),
                     Base=rep(c("A","C","G","T"),each=4))
```

---

# Assessment by cycle  with raw ChIPseq sequence data.

Now we can plot the frequencies using ggplot2


```r
ggplot(toPlot,aes(y=Count,x=Cycle,colour=Base))+geom_line()+
  theme_bw()
```

![](ChIPseq_In_Bioconductor_files/figure-html/mycRep1ReadsAlpByCyclePlot3-1.png)&lt;!-- --&gt;
---

# Assessment by cycle  with raw ChIPseq sequence data.

We can also assess mean read quality over cycles. This will allow us to identify whether there are any isses with quality dropping off over time.

For this we use the [**as(*read_quality*,"matrix")**](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#27) function first to translate our ASCI quality scores to numeric quality scores.


```r
qualAsMatrix &lt;- as(readQuality,"matrix")
qualAsMatrix[1:2,]
```

```
##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
## [1,]   31    2    2    2    2    2    2    2    2     2     2     2     2
## [2,]   31   33   33   34   33   33   25   31   28    33    33    33    33
##      [,14] [,15] [,16] [,17] [,18] [,19] [,20] [,21] [,22] [,23] [,24]
## [1,]     2     2     2     2     2     2     2     2     2     2     2
## [2,]    34    32    33    33    32    33    33    33    33    33    34
##      [,25] [,26] [,27] [,28] [,29] [,30] [,31] [,32] [,33] [,34] [,35]
## [1,]     2     2     2     2     2     2     2     2     2     2     2
## [2,]    32    33    30    33    33    33    33    32    33    32    33
##      [,36]
## [1,]     2
## [2,]    33
```

---

# Assessment by cycle  with raw ChIPseq sequence data.

We can now [visualise qualities across cycles using a boxplot.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/answers/fastq_answers.html)


```r
boxplot(qualAsMatrix[1:1000,])
```

![](ChIPseq_In_Bioconductor_files/figure-html/mycRep1ReadsQByCyclePlot-1.png)&lt;!-- --&gt;

---

# Assessment by cycle  with raw ChIPseq sequence data.

In this case the distribution of reads quality scores and read qualities over time look okay. We will often want to access fastQ samples together to see if any samples stick out by these metrics.

Here we observed a second population of low quality scores so will remove some reads with low quality scores and high unknown bases.

---

# Filtering data

We will want to conserve our memory usage to [allow us to deal with loading large files.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#40)

Here we set up a **FastqStreamer object** to read in 100000 reads at a time. 


```r
fqStreamer &lt;- FastqStreamer("~/Downloads/ENCFF001NQP.fastq.gz",
                            n=100000)
```

---

# Filtering data

Now we [loop through file](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#41), filter reads and [write out a FastQ](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/FastQInBioconductor.html#37) of our filtered reads


```r
TotalReads &lt;- 0
TotalReadsFilt &lt;- 0
while (length(fq &lt;- yield(fqStreamer))&gt;0) {
    TotalReads &lt;- TotalReads+length(fq)
    filt1 &lt;- fq[alphabetScore(fq) &gt; 300 ]
    filt2 &lt;- filt1[alphabetFrequency(sread(filt1))[,"N"] &lt; 10]
    TotalReadsFilt &lt;- TotalReadsFilt+length(filt2)
    writeFastq(filt2,"filtered_ENCFF001NQP.fastq.gz",mode="a")
}
TotalReads
TotalReadsFilt
```

---
# Aligning ChIPseq reads

Following assessment of read quality and any read filtering we applied, we will want to align our reads to the genome so as to identify any genomic locations showing enrichment for aligned reads above background.

Since ChIPseq reads will align continously agaist our reference genome we can use [our genomic aligners we have seen in previous sessions.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignmentInBioconductor.html#6) The resulting BAM file will contain aligned sequence reads for use in further analysis.

&lt;div align="center"&gt;
&lt;img src="imgs/sam.png" alt="igv" height="300" width="500"&gt;
&lt;/div&gt;

---
# Creating a reference genome.

First we need to retrieve the sequence information for the genome of interest in [FASTA format](https://rockefelleruniversity.github.io/Genomic_Data/genomicFileFormats.html#/6)

We can use the [BSgenome libraries to retrieve the full sequence information.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/SequencesInBioconductor.html#4)

For the mouse mm10 genome we load the package **BSgenome.Mmusculus.UCSC.mm10**.




```r
library(BSgenome.Mmusculus.UCSC.mm10)
BSgenome.Mmusculus.UCSC.mm10
```

```
## Mouse genome:
## # organism: Mus musculus (Mouse)
## # provider: UCSC
## # provider version: mm10
## # release date: Dec. 2011
## # release name: Genome Reference Consortium GRCm38
## # 66 sequences:
## #   chr1                 chr2                 chr3                
## #   chr4                 chr5                 chr6                
## #   chr7                 chr8                 chr9                
## #   chr10                chr11                chr12               
## #   chr13                chr14                chr15               
## #   ...                  ...                  ...                 
## #   chrUn_GL456372       chrUn_GL456378       chrUn_GL456379      
## #   chrUn_GL456381       chrUn_GL456382       chrUn_GL456383      
## #   chrUn_GL456385       chrUn_GL456387       chrUn_GL456389      
## #   chrUn_GL456390       chrUn_GL456392       chrUn_GL456393      
## #   chrUn_GL456394       chrUn_GL456396       chrUn_JH584304      
## # (use 'seqnames()' to see all the sequence names, use the '$' or '[['
## # operator to access a given sequence)
```

---
# Creating a reference genome.

We will only use the major chromosomes for our analysis so we may exclude random and unplaced contigs.
Here we cycle through the major chromosomes and create a [**DNAStringSet** object from the retrieved sequences](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/SequencesInBioconductor.html#17).


```r
mainChromosomes &lt;- paste0("chr",c(1:19,"X","Y","M"))
mainChrSeq &lt;- lapply(mainChromosomes,
                     function(x)BSgenome.Mmusculus.UCSC.mm10[[x]])
names(mainChrSeq) &lt;- mainChromosomes
mainChrSeqSet &lt;- DNAStringSet(mainChrSeq)
mainChrSeqSet
```

```
##   A DNAStringSet instance of length 22
##          width seq                                     names               
##  [1] 195471971 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr1
##  [2] 182113224 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr2
##  [3] 160039680 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr3
##  [4] 156508116 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr4
##  [5] 151834684 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr5
##  ...       ... ...
## [18]  90702639 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr18
## [19]  61431566 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chr19
## [20] 171031299 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chrX
## [21]  91744698 NNNNNNNNNNNNNNNNNN...NNNNNNNNNNNNNNNNNN chrY
## [22]     16299 GTTAATGTAGCTTAATAA...CGCAATAAACATTAACAA chrM
```

---
# Creating a reference genome.

Now we have a **DNAStringSet** object we can use the [**writeXStringSet** to create our FASTA file of sequences to align to.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/SequencesInBioconductor.html#17)


```r
writeXStringSet(mainChrSeqSet,
                "BSgenome.Mmusculus.UCSC.mm10.mainChrs.fa")
```


---
# QuasR ChIP-seq alignment - Sample table

We can align our raw sequence data in fastQ format to the new FASTA file of our mm10 genome sequence using the **QuasR** package.

As we have seen in previous sessions, [the QuasR package requires a sample table to specify alignment files.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignmentInBioconductor.html#15)

The sample table requires is a tab-delimited file listing the path to fastq(s) to be aligned and the desired sample names.


```r
FileName &lt;- "filtered_ENCFF001NQP.fastq.gz"
SampleName &lt;- "Myc_Mel_1"
sampleTable &lt;- data.frame(FileName,SampleName)
write.table(sampleTable,file="sampleTable.txt",sep="\t",
            quote=FALSE,row.names = FALSE)
```


```
##                        FileName SampleName
## 1 filtered_ENCFF001NQP.fastq.gz  Myc_Mel_1
```

---
# QuasR genomic with BSGenome object

We can now align our data by using the **qAlign()** function.

The **qAlign()** function requires the path to our sample table file and the path of a FASTA file to align sequence reads against.


```r
library(QuasR)
qAlign("sampleTable.txt","BSgenome.Mmusculus.UCSC.mm10.mainChrs.fa")
```

---

# Aligning ChIPseq reads with Rsubread.

The **Rsubread** package offers a faster aligner than the QuasR package although the **Rsubread** package is only available on Mac OS and Linux.

For alignment with the **Rsubread** package we must first build our genome index for Rsubread using the **buildindex()** function.

The [**buildindex()** function simply takes the parameters of our desired index name and the FASTA file to build index from.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignmentInBioconductor.html#34)



```r
library(Rsubread)
buildindex("BSgenome.Mmusculus.UCSC.mm10.mainChrs",
           "BSgenome.Mmusculus.UCSC.mm10.mainChrs.fa")
```

---
# Aligning ChIPseq reads with Rsubread.

Having constructed our index, we can align our reads using the **Rsubread** package's **align()** function.

The [**align()** function accepts arguments for the index to align to, the fastQ to align, the name of output BAM, the mode of alignment (rna or dna) and the phredOffset.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignmentInBioconductor.html#34)

Note that here we set the phredOffset to be 64. Rsubread will provide a warning if we set this wrong.


```r
align("BSgenome.Mmusculus.UCSC.mm10.mainChrs",
      "filtered_ENCFF001NQP.fastq.gz",
      output_file="filtered_ENCFF001NQP.bam",
      type="dna",phredOffset = 64)
```


---

# Sort and index reads.

As before, we sort and index our files using the [**Rsamtools** packages **sortBam()** and **indexBam()** functions respectively.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignedDataInBioconductor.html#10)

The resulting sorted and indexed BAM file is now ready for use in external programs such as IGV as well as for further downstream analysis in R.


```r
library(Rsamtools)
sortBam("filtered_ENCFF001NQP.bam","SR_Myc_Mel_rep1")
indexBam("SR_Myc_Mel_rep1.bam")
```
---

# Mapped reads.

Now we have the index for the BAM file, we can retrieve and plot the number of mapped reads using [the **idxstatsBam()** function.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignedDataInBioconductor.html#15)


```r
mappedReads &lt;- idxstatsBam("../../../SR_Myc_Mel_rep1.bam")
TotalMapped &lt;- sum(mappedReads[,"mapped"])
ggplot(mappedReads,aes(x=seqnames,y=mapped))+
  geom_bar(stat="identity")
```

![](ChIPseq_In_Bioconductor_files/figure-html/mapped-1.png)&lt;!-- --&gt;

---

# Create a bigWig

We can also create a bigWig from our sorted, indexed BAM file to allow us to quickly review our data in IGV.

First we use the [**coverage()** function to create an **RLElist object** containing our coverage scores.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/Summarising_Scores_In_Bioconductor.html#13)


```r
forBigWig &lt;- coverage("SR_Myc_Mel_rep1.bam")
forBigWig
```

---

# Create a bigWig

We can now export our [**RLElist object** as a bigWig using the **rtracklayer** package's **export.bw()** function.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicScores_In_Bioconductor.html#40)


```r
library(rtracklayer)
export.bw(forBigWig,con="SR_Myc_Mel_rep1.bw")
```


---
# Create a bigWig

We may wish to normalise our coverage to allow us to compare enrichment across samples.

We can use the [**weight** parameter in the **coverage()**](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/Summarising_Scores_In_Bioconductor.html#19) to scale our reads to the number of mapped reads multiplied by a million (reads per million).


```r
forBigWig &lt;- coverage("SR_Myc_Mel_rep1.bam",
                      weight = (10^6)/TotalMapped)
forBigWig
export.bw(forBigWig,con="SR_Myc_Mel_rep1_weighted.bw")
```

---
# BAM and bigWig

![](imgs/chipBw.png)

---

# Time for an exercise.

[Link_to_exercises](../../Exercises/chipseq_part1_exercises.html)

[Link_to_answers](../../Answers/chipseq_part1_answers.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
